### PR TITLE
[RLlib] Checkpoint and restore connectors.

### PR DESCRIPTION
**Pull request by gjoliver**
Plus a couple of examples showing the usage of connector enabled policies.

## Why are these changes needed?

Allow checkpoint and restore of connector pipelines.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [*] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/connectors/connector.py [hunk 1]
```diff
@@ -365,6 +377,11 @@ def prepend(self, connector: Connector):
         """
         self.connectors.insert(0, connector)
 
+        print(
```
Comment (sven1977) [inline rllib/connectors/connector.py:380]: same

File: rllib/connectors/connector.py [hunk 2]
```diff
@@ -373,6 +390,11 @@ def append(self, connector: Connector):
         """
         self.connectors.append(connector)
 
+        print(
```
Comment (sven1977) [inline rllib/connectors/connector.py:393]: same

File: rllib/examples/connectors/adapt_connector_policy.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
+"""This example script shows how to load a connector enabled policy,
+and adapt/use it with a different version of environment.
```
Comment (sven1977) [inline rllib/examples/connectors/adapt_connector_policy.py:2]: nit: "the environment"

File: rllib/policy/policy.py [hunk 4]
```diff
@@ -740,8 +740,44 @@ def get_state(self) -> PolicyState:
             # The current global timestep.
             "global_timestep": self.global_timestep,
         }
+        if self.config.get("enable_connectors", False):
```
Comment (sven1977) [inline rllib/policy/policy.py:743]: Can you add a one-line comment here? that we are adding the connector state?

File: rllib/policy/policy.py [hunk 5]
```diff
@@ -740,8 +740,44 @@ def get_state(self) -> PolicyState:
...
+            connector_configs = {}
+            if self.agent_connectors:
+                connector_configs["agent"] = self.agent_connectors.to_config()
+            if self.action_connectors:
+                connector_configs["action"] = self.action_connectors.to_config()
+            state["connector_configs"] = connector_configs
         return state
 
+    @ExperimentalAPI
```
Comment (sven1977) [inline rllib/policy/policy.py:752]: Should we use `PublicApi(alpha)` here as well?

File: rllib/policy/policy.py [hunk 6]
```diff
@@ -740,8 +740,44 @@ def get_state(self) -> PolicyState:
...
+    def restore_connectors(self, state: PolicyState):
+        """Restore agent and action connectors if configs available.
+
+        Args:
+            state: The new state to set this policy to. Can be
+                obtained by calling `self.get_state()`.
+        """
+        # To avoid a circular dependency problem cause by SampleBatch.
```
Comment (sven1977) [inline rllib/policy/policy.py:760]: Future note to ourselves: We should move SampleBatch out of the policy folder. It doesn't belong there.

File: rllib/policy/policy.py [hunk 7]
```diff
@@ -740,8 +740,44 @@ def get_state(self) -> PolicyState:
...
+
+        if not self.config.get("enable_connectors", False):
+            return
+
+        connector_configs = state.get("connector_configs", {})
+        if "agent" in connector_configs:
+            self.agent_connectors = restore_connectors_for_policy(
+                self, connector_configs["agent"]
+            )
+            print("restoring agent connectors:")
```
Comment (sven1977) [inline rllib/policy/policy.py:771]: logger.info

File: rllib/policy/policy_map.py [hunk 8]
```diff
@@ -118,48 +118,18 @@ def create_policy(
...
-            else:
-                with tf1.variable_scope(var_scope):
-                    self[policy_id] = class_(
-                        observation_space, action_space, merged_config
-                    )
-        # Non-tf: No graph, no session.
-        else:
-            class_ = policy_cls
-            self[policy_id] = class_(observation_space, action_space, merged_config)
+        _class = get_tf_eager_cls_if_necessary(policy_cls, merged_config)
```
Comment (sven1977) [inline rllib/policy/policy_map.py:121]: nice! thanks for cleaning this up and creating the utility function

**Review by sven1977 — APPROVED** [review]
Awesome PR @gjoliver , thanks for the examples on connectors.
Just a few nits, then we can merge.

Oh, one last thing, we should add the new examples to BUILD!

File: rllib/examples/connectors/adapt_connector_policy.py (hunk 3 above)
Comment (gjoliver) [reply rllib/examples/connectors/adapt_connector_policy.py:2]: done

File: rllib/examples/connectors/run_connector_policy.py [hunk 9]
```diff
@@ -0,0 +1,54 @@
+"""This example script shows how to load a connector enabled policy,
+and adapt and use it with a different version of environment.
```
[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/examples/connectors/run_connector_policy.py:2]: updated the comment.
thanks.

File: rllib/policy/policy.py (hunk 4 above)
Comment (gjoliver) [reply rllib/policy/policy.py:743]: sure

File: rllib/policy/policy.py (hunk 5 above)
Comment (gjoliver) [reply rllib/policy/policy.py:752]: 👍

File: rllib/policy/policy.py (hunk 6 above)
Comment (gjoliver) [reply rllib/policy/policy.py:760]: yeah, totally, SampleBatch should be a pretty low level util that depends on nothing.

File: rllib/policy/policy_map.py (hunk 8 above)
Comment (gjoliver) [reply rllib/policy/policy_map.py:121]: 👌

File: rllib/policy/policy.py (hunk 7 above)
Comment (gjoliver) [reply rllib/policy/policy.py:771]: done

File: rllib/policy/policy.py [hunk 10]
```diff
@@ -740,8 +740,44 @@ def get_state(self) -> PolicyState:
...
+            print(self.agent_connectors.__str__(indentation=4))
+        if "action" in connector_configs:
+            self.action_connectors = restore_connectors_for_policy(
+                self, connector_configs["action"]
+            )
+            print("restoring action connectors:")
+            print(self.action_connectors.__str__(indentation=4))
```
[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/policy/policy.py:778]: done

File: rllib/connectors/connector.py (hunk 2 above)
Comment (gjoliver) [reply rllib/connectors/connector.py:393]: done

File: rllib/connectors/connector.py (hunk 1 above)
Comment (gjoliver) [reply rllib/connectors/connector.py:380]: done

File: rllib/connectors/connector.py [hunk 11]
```diff
@@ -357,6 +364,11 @@ def insert_after(self, name: str, connector: Connector):
             raise ValueError(f"Can not find connector {name}")
         self.connectors.insert(idx + 1, connector)
 
+        print(
```
[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/connectors/connector.py:367]: done

File: rllib/connectors/connector.py [hunk 12]
```diff
@@ -341,6 +343,11 @@ def insert_before(self, name: str, connector: Connector):
             raise ValueError(f"Can not find connector {name}")
         self.connectors.insert(idx, connector)
 
+        print(
```
[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/connectors/connector.py:346]: done

**Comment by gjoliver** [comment]
ok, addressed all the comments. let's see if CI is happy.
completely understand your comments about logger.info(), was a bit hesitant because you often can't see log files on disks on Anyscale :).

Also added all the examples as unit tests.
I also added a multi-agent example, where we restore a TF PPO policy to train a new Torch SAC policy.

**Pull request merged**